ISIS-adjacency polling support

ottorei opened the pull request:
Adds a very basic ISIS-protocol support with the functionality to poll configured adjacencies. This information may be used to trigger alarms when the device loses any of the configured adjacencies. Also adds a simple table for device and routing contexts that can be used to view the general status of the topology. Also, links are created on the UI where possible - requires that the device's ifIndexes are saved into the DB.

Each polled adjacency is saved into the DB during the first time it is seen. When polled again, the old record is matched and updated by local ifIndex + device_id as a unique identifier. If the port has ISIS configuration but there has never been any adjacents, this saves only the ifIndex and marks it's status as down.

==> Tested with Juniper MX-80, MX-240 and Cisco ASR9K-series routers. Seems to poll all configured adjacencies

- [x] Basic polling support (== Adjacencies and their status can be polled to some extent)
- [x] Adjacencies can be saved into the DB
- [x] Basic tables in device and routing menus
- [x] Module can be enabled from polling options
- [x] Translations for settings
- [x] Alert fault listing (functions.inc.php) by adjacency IP - could also generate links if the adjacency is found in NMS
- [x] Routing menubar link
- [x] Add alert template for adjacency status check
- [x] Device menubar link
- [x] Cleanup options from poller module (missing as in extra entries that exist in the DB but do not exist on the poll. These entries should be purged from the DB)
- [x] Code style fixes
- [x] Generate links on routing page pages where possible (point interfaces and / or IPs to the devices in question if they exists on the NMS)
- [x] Remove entries from the DB when the device is deleted
- [x] Data formatting
- [x] Database schema and migrations
- [x] Convert pollert module to new style
- [x] Test cases

#### Testing on your instance: 
This patch can be tested with : 
`./scripts/github-apply 12461 `

#### Screenshots
Menu entry on main page:
![isis_routing_menu](https://user-images.githubusercontent.com/7031571/105777539-37625100-5f73-11eb-9d8d-5a624abba023.PNG)

Routing menu from main page:
![isis_routing_view](https://user-images.githubusercontent.com/7031571/105777536-36312400-5f73-11eb-869b-c20aa664ddb4.png)

Routing entries on device page:
![isis_device_page](https://user-images.githubusercontent.com/7031571/105777540-37625100-5f73-11eb-8bbe-4f15ee321d2a.png)


DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 12461 `
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

Jellyfrog commented on app/Http/Controllers/Device/Tabs/RoutingController.php:
```diff
@@ -38,6 +39,7 @@ public function __construct()
         $device = DeviceCache::getPrimary();
         $this->tabs = [
             'ospf' => $device->ospfInstances()->count(),
+            'isis' => IsisAdjacency::where('device_id', $device->device_id)->count(),
```
use relation instead

Jellyfrog commented on database/migrations/2021_25_01_0127_create_isis_adjacencies_table.php:
```diff
@@ -0,0 +1,39 @@
+    {
+        Schema::create('isis_adjacencies', function (Blueprint $table) {
+            $table->increments('id');
+            $table->integer('device_id');
+            $table->integer('port_id');
```
missing index on both

ottorei replied:
$table->index(['device_id', 'port_id']) should do this?

Jellyfrog commented on includes/html/pages/device/routing/isis.inc.php:
```diff
@@ -0,0 +1,57 @@
+        $color = 'red';
+    }
+
+    $interface_name = Port::query()->where('port_id', $adj->port_id)->first()->ifName;
```
creating N+1 queries, use `with()` on your IsisAdjacency query.

murrant replied:
`IsisAdjacency::where('device_id', $device['device_id'])->with('port')->get()`

and then `$adj->port->ifName`

ottorei replied:
Fixed

Jellyfrog commented on includes/html/pages/routing/isis.inc.php:
```diff
@@ -0,0 +1,62 @@
+        $color = 'red';
+    }
+
+    $interface_name = Port::query()->where('port_id', $adj->port_id)->first()->ifName;
```
same as above

Jellyfrog commented on includes/polling/isis.inc.php:
```diff
@@ -0,0 +1,91 @@
+    $adjacencies,
+    $isis_data,
+    $tmp_adjacencies,
+    $tmp_time,
```
```suggestion
    $tmp_time
```

murrant commented on includes/functions.php:
```diff
@@ -353,6 +353,10 @@ function delete_device($id)
 
+    //Remove IsisAdjacencies
+    \App\Models\IsisAdjacency::where('device_id', $id)->delete();
+    \App\Models\IsisAdjacency::where('device_id', $id)->delete();
```
Duplicate line

ottorei replied:
Removed the other one

murrant commented on includes/html/functions.inc.php:
```diff
@@ -936,6 +936,17 @@ function alert_details($details)
             $fallback = false;
         }
 
+        if ($tmp_alerts['isisISAdjIPAddrAddress']) {
```
What is the reason for this special case?

ottorei replied:
To format the alert to show only the adjacents and not general polling time or device's uptime when expanded. See the screenshot.
![isis_special_case](https://user-images.githubusercontent.com/7031571/106005504-f8401700-60bc-11eb-819c-420a7cccf829.PNG)

murrant replied:
Hmm, this feels like the wrong place for that, but I don't have any help beyond that, sorry.

murrant replied:
I suggest  moving this code to a separate PR, it is going to be a blocker for this PR.

ottorei replied:
OK, I will remove this and make a separate PR for this later.

murrant commented on includes/polling/isis.inc.php:
```diff
@@ -0,0 +1,91 @@
+<?php
```
I would prefer if this were a new style poller module.  See the modules in `LibreNMS/Modules`

ottorei replied:
Changed to use the new style

ottorei reviewed (commented):
Review changes

murrant commented:
Also one more thing needed for all new modules is testing support.  You will need to add an entry to `tests/module_tables.yaml` then capture and generate test data for at least 1 device.

ottorei commented:
Basic operations are working now. The schema needs to be changed to also save the ifIndex as a fallback in case the port_id can not be found for example due to filtering.

murrant commented:
I'm not sure handling when the port is missing is a worthwhile case to code around.   Maybe send an event log when that occurs is enough.

ottorei commented:
> 
> 
> I'm not sure handling when the port is missing is a worthwhile case to code around. Maybe send an event log when that occurs is enough.

I was thinking this too at first but after testing with lab gear I found that there might be some cases where the port_id can not be found. This is particularly an issue with Juniper gear with SNMP-filters. By also saving the ifIndex the adjacencies can be reliably linked to each device in these cases. This problem only occurs if there is no port data of the device's ports available.

murrant commented on misc/config_definitions.json:
```diff
@@ -4286,6 +4286,13 @@
             "type": "boolean"
         },
+        "poller_modules.isis": {
+            "order": 331,
```
331 does not seem alphabetical.

murrant commented on LibreNMS/Modules/Isis.php:
```diff
@@ -0,0 +1,164 @@
+        }
+
+        // Loop through all configured adjacencies on the device
+        foreach ($circuits_poll as $circuit => $circuit_data) {
```
Instead of this huge foreach loop, take a look at using the SyncsModels trait.  See https://github.com/librenms/librenms/blob/master/LibreNMS/Modules/Mempools.php#L57 for an example.

ottorei replied:
All circuits would still need to be polled every time since the adjacency is lost when the configured adjacency goes down for a while. The only way to detect these is to first poll all configured circuits and then poll all adjacencies for these circuits. The circuit will always return the correct value if there is an ISIS-configuration on it but the adjacencies table does only return values if the adjacency is up at the moment the polling is done. That is why I am saving all values and updating them based on the ifIndex of the device. That way the missing and downed adjacencies can be updated and they also retaing useful information such as IfIndex and system id while they are down.

Jellyfrog replied:
After implementing SyncsModels in https://github.com/librenms/librenms/pull/12605, i can totally recommend it ^^

murrant commented on LibreNMS/Modules/Isis.php:
```diff
@@ -0,0 +1,164 @@
+        $device_id = $os->getDeviceId();
+
+        // Poll all ISIS enabled circuits from the device
+        $circuits_poll = $os->getCacheTable('ISIS-MIB::isisCirc', 'ISIS-MIB');
```
getCacheTable() makes the data available outside this module by storing it on the OS.  Is that needed?

ottorei replied:
Good point, that is not actually needed.

murrant commented:
@ottorei this is failing because of an upstream change.  You can either update your branch and re-capture test data or update the json files by hand.

PipoCanaja commented on mibs/junos/ISIS-MIB:
```diff
@@ -0,0 +1,4581 @@
+ISIS-MIB DEFINITIONS ::= BEGIN
```
Hello
ISIS-MIB is a standard MIB, it must be located in the mibs folder, not in the ,mibs/junos folder. And in fact, is is already in the mibs folder, and seems to be the same version. Please check and remove the mibs/junos one.
Thanx

ottorei replied:
For some reason the ISIS-polling does not seem to work at all when this file is removed. Do I need to specify the location for those generic MIBs?

ottorei replied:
Found the issue - there seems to be an old MIB-file with possible incorrect OIDs in the junos-folder. Deleting that seems to resolve this issue.

PipoCanaja reviewed (changes requested):
Mib file

Jellyfrog commented on LibreNMS/Modules/Isis.php:
```diff
@@ -0,0 +1,166 @@
+                $port_id = (int) $device->ports()->where('ifIndex', $circuit)->value('port_id');
+
+                if ($circuit_data['isisCircPassiveCircuit'] != '1') {
+                    //var_dump($adjacencies_poll[$circuit]['isisISAdjState']);
```
Debug code

ottorei replied:
Removed debug code

Jellyfrog commented on includes/html/pages/device/routing/isis.inc.php:
```diff
@@ -0,0 +1,63 @@
+
+    //dd($adj);
+    $interface_name = $adj->port->ifName;
+    //dd($interface_name);
```
Debug code

Jellyfrog commented on includes/html/pages/device/routing/isis.inc.php:
```diff
@@ -0,0 +1,63 @@
+          </tr>
+        </thead>';
+
+// FIXME: Can refer to ports, production.ERROR: Call to undefined relationship [Port] on model [App\Models\IsisAdjacency]
```
Still relevant?

ottorei replied:
Not anymore, I'll push updates probably tomorrow.

ottorei replied:
Removed old commend

murrant commented on LibreNMS/Modules/Isis.php:
```diff
@@ -0,0 +1,166 @@
+        $isis_data = [];
+
+        // Translate system state codes into meaningful strings
+        $isis_codes['1'] = 'L1';
```
Please initialize $isis_codes.  The best way is to just set it to a fully formed array.

ottorei replied:
Initialized

ottorei reviewed (commented):
Deleted debug code and unnecessary mib-file

murrant commented:
Just a little more cleanup from the lint failures.

murrant commented:
Can you have discovery check `ISIS-MIB::isisSysAdminState` or something, then store that as a component or device attribute and check it in  the poller to see if we need to poll ISIS.  That way we can enable this for all OS.

Instead of end() you could use \Illuminate\Support\Arr::last()

ottorei commented:
> 
> 
> Can you have discovery check `ISIS-MIB::isisSysAdminState` or something, then store that as a component or device attribute and check it in the poller to see if we need to poll ISIS. That way we can enable this for all OS.
> 
> Instead of end() you could use \Illuminate\Support\Arr::last()

Where should this information be stored? Is there already a place for this?

murrant commented:
> > Can you have discovery check `ISIS-MIB::isisSysAdminState` or something, then store that as a component or device attribute and check it in the poller to see if we need to poll ISIS. That way we can enable this for all OS.
> > Instead of end() you could use \Illuminate\Support\Arr::last()
> 
> Where should this information be stored? Is there already a place for this?

I gave you two options  `Component` or `DeviceAttrib`.

ottorei commented:
> 
> 
> Can you have discovery check `ISIS-MIB::isisSysAdminState` or something, then store that as a component or device attribute and check it in the poller to see if we need to poll ISIS. That way we can enable this for all OS.
> 
> Instead of end() you could use \Illuminate\Support\Arr::last()

I added a discovery process for adding this as a component

> 
> 
> > > Can you have discovery check `ISIS-MIB::isisSysAdminState` or something, then store that as a component or device attribute and check it in the poller to see if we need to poll ISIS. That way we can enable this for all OS.
> > > Instead of end() you could use \Illuminate\Support\Arr::last()
> > 
> > 
> > Where should this information be stored? Is there already a place for this?
> 
> I gave you two options `Component` or `DeviceAttrib`.

Added a simple component check.
